### PR TITLE
Close #13624: Refactor ViewportInteractionItem to use strong enum

### LINKED ISSUE #13624: Refactor ViewportInteractionItem to use strong enum

### PR DESCRIPTION
**Pull request by ddevrien**
Refactored ViewportInteractionItem (#13624). clang-format seems to fail but I'm not sure why. Any help would be appreciated, I have very little experience with contributions.

### CODE REVIEW

File: src/openrct2/interface/Viewport.h [hunk 1]
```diff
@@ -49,38 +49,37 @@ enum
...
-    VIEWPORT_INTERACTION_ITEM_FOOTPATH_ITEM,
-    VIEWPORT_INTERACTION_ITEM_PARK,
-    VIEWPORT_INTERACTION_ITEM_WALL,
-    VIEWPORT_INTERACTION_ITEM_LARGE_SCENERY,
-    VIEWPORT_INTERACTION_ITEM_LABEL,
-    VIEWPORT_INTERACTION_ITEM_BANNER,
-
+    None,
+    Terrain,
+    Sprite,
```
Comment (duncanspumpkin) [inline src/openrct2/interface/Viewport.h:56]: Entity is the preferred name for this these days

File: src/openrct2/interface/Viewport.h [hunk 2]
```diff
@@ -49,38 +49,37 @@ enum
...
+    Ride,
+    Water,
+    Scenery,
+    Footpath,
+    FootpathItem,
+    Park,
```
Comment (duncanspumpkin) [inline src/openrct2/interface/Viewport.h:62]: ParkEntrance

**Comment by ddevrien** [comment]
processed comments + fixed clang-format. The build is still failing though, caused by a bunch of these errors: 
`error: enumeration value ‘Terrain’ not handled in switch [-Werror=switch]`
How should I handle these?

**Comment by ddevrien** [comment]
Candidate for squash merge :D. I'm learning along the way, it took a few commits to get through all the checks.

File: src/openrct2/interface/Viewport.cpp [hunk 3]
```diff
@@ -1277,17 +1277,16 @@ void viewport_set_visibility(uint8_t mode)
...
+    if (ps->sprite_type == ViewportInteractionItem::None || ps->sprite_type == ViewportInteractionItem::Label
+        || ps->sprite_type > ViewportInteractionItem::Banner)
         return false;
 
     uint16_t mask;
-    if (ps->sprite_type == VIEWPORT_INTERACTION_ITEM_BANNER)
+    if (ps->sprite_type == ViewportInteractionItem::Banner)
         // I think CS made a typo here. Let's replicate the original behaviour.
-        mask = 1 << (ps->sprite_type - 3);
+        mask = 1 << (static_cast<uint8_t>(ps->sprite_type) - 3);
```
Comment (Gymnasiast) [inline src/openrct2/interface/Viewport.cpp:1287]: We have `EnumValue` for this kind of thing, which look a bit nicer.

File: src/openrct2/interface/Viewport.h [hunk 4]
```diff
@@ -49,38 +49,38 @@ enum
...
-    VIEWPORT_INTERACTION_MASK_RIDE = ~(1 << (VIEWPORT_INTERACTION_ITEM_RIDE - 1)),
-    VIEWPORT_INTERACTION_MASK_WATER = ~(1 << (VIEWPORT_INTERACTION_ITEM_WATER - 1)),
-    VIEWPORT_INTERACTION_MASK_SCENERY = ~(1 << (VIEWPORT_INTERACTION_ITEM_SCENERY - 1)),
-    VIEWPORT_INTERACTION_MASK_FOOTPATH = ~(1 << (VIEWPORT_INTERACTION_ITEM_FOOTPATH - 1)),
-    VIEWPORT_INTERACTION_MASK_FOOTPATH_ITEM = ~(1 << (VIEWPORT_INTERACTION_ITEM_FOOTPATH_ITEM - 1)),
-    VIEWPORT_INTERACTION_MASK_PARK = ~(1 << (VIEWPORT_INTERACTION_ITEM_PARK - 1)),
-    VIEWPORT_INTERACTION_MASK_WALL = ~(1 << (VIEWPORT_INTERACTION_ITEM_WALL - 1)),
-    VIEWPORT_INTERACTION_MASK_LARGE_SCENERY = ~(1 << (VIEWPORT_INTERACTION_ITEM_LARGE_SCENERY - 1)),
-    VIEWPORT_INTERACTION_MASK_BANNER = ~(1 << (VIEWPORT_INTERACTION_ITEM_BANNER - 2)), // Note the -2 for BANNER
+    VIEWPORT_INTERACTION_MASK_TERRAIN = ~(1 << (static_cast<uint8_t>(ViewportInteractionItem::Terrain) - 1)),
```
Comment (Gymnasiast) [inline src/openrct2/interface/Viewport.h:72]: `EnumValue`

File: src/openrct2/interface/Viewport.cpp [hunk 5]
```diff
@@ -1277,17 +1277,16 @@ void viewport_set_visibility(uint8_t mode)
  */
 static bool PSSpriteTypeIsInFilter(paint_struct* ps, uint16_t filter)
 {
-    if (ps->sprite_type == VIEWPORT_INTERACTION_ITEM_NONE
-        || ps->sprite_type == 11 // 11 as a type seems to not exist, maybe part of the typo mentioned later on.
```
Comment (tupaschoal) [inline src/openrct2/interface/Viewport.cpp:1281]: This code was a bit weird before (the `== 11`), it looks more correct now, but I'm commenting to see what other people think

File: test/testpaint/TestPaint.cpp [hunk 6]
```diff
@@ -27,8 +27,8 @@ namespace TestPaint
 {
     void ResetEnvironment()
     {
-        gPaintInteractionType = VIEWPORT_INTERACTION_ITEM_RIDE;
-        gPaintSession.InteractionType = VIEWPORT_INTERACTION_ITEM_RIDE;
+        gPaintInteractionType = static_cast<uint8_t>(ViewportInteractionItem::Ride);
```
Comment (tupaschoal) [inline test/testpaint/TestPaint.cpp:30]:
```suggestion
        gPaintInteractionType = EnumValue(ViewportInteractionItem::Ride);
```

File: test/testpaint/TestPaint.cpp [hunk 7]
```diff
@@ -27,8 +27,8 @@ namespace TestPaint
...
+        gPaintSession.InteractionType = ViewportInteractionItem::Ride;
```
Comment (tupaschoal) [inline test/testpaint/TestPaint.cpp:31]: These two seem unused 🤔

Comment (duncanspumpkin) [reply test/testpaint/TestPaint.cpp:31]: I would leave for now.

File: src/openrct2/interface/Viewport.cpp (hunk 5 above)
Comment (duncanspumpkin) [reply src/openrct2/interface/Viewport.cpp:1281]: This is checking for the types that can't be interacted with. So thats: None, labels and anything invalid (> banner)

File: test/testpaint/TestPaint.cpp (hunk 7 above)
Comment (ddevrien) [reply test/testpaint/TestPaint.cpp:31]: I'm not sure what those tests are about, it seems only Linux (i686, portable) is running testpaint. I'll leave it in but change the static cast to EnumValue as suggested.

Comment (duncanspumpkin) [reply test/testpaint/TestPaint.cpp]: Its meant to check that there have been no changes to the painting code for all rides compared to the original game (hence why its only x86 compatible). It doesn't quite work correctly at the moment due to other changes but I'm trying to get it running again on all platforms and in the process this part will be removed so don't worry too much about making this good.

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**